pyigm/clustering/xcorr_utils.py

```diff
@@ -222,3 +225,136 @@ def clean_matrix(W,n=-99):
+    is calculated from at least Nmin (=20) galaxies (i.e. the magnitude
+    band is increased).
+
+    Parameters
+    ----------
+    galreal
+    Nrand
+    Nmin : int, optional
+    DZ : float, optional
+      delta z for the histogram for getting the
+    smooth_scale : float, optional
+       smoothing scale for the histogram (in number of bins, so depends on DZ)
+
+    Returns
+    -------
+    galrand : np rec array
+      Copy of galreal, Nrand times
+    """
+    from pyigm.clustering.randist import RanDist
+    from scipy.interpolate import CubicSpline
+    from scipy.ndimage import gaussian_filter as gf
+
+    debug = 0
+    Ckms = 299792.458
+    Nmin = int(Nmin)  # minimum number of galaxys for the fit
+    zmin = np.max([np.min(galreal.ZGAL[galreal.ZGAL > 0]), 1e-9])
+    zmax = np.max(galreal.ZGAL)
+    # spline in z
+    galreal.sort(order='MAG')  # np.recarray.sort()
+    galrand = galreal.repeat(Nrand)
+    delta_mag = 0.5
+
+    bins = np.append(np.linspace(0, zmin, 20), np.arange(zmin + DZ, zmax + 10 * DZ, DZ))
+    # bins = np.arange(0, zmax+DZ, DZ)
+
+    # Make subhistograms depending on magnitude. Use dictionary.
+    VALS = dict()
+    SPL = dict()
+    aux_hist, _ = np.histogram(galreal.ZGAL, bins)
+    VALS['all'] = gf(aux_hist.astype(float), smooth_scale)  # smooth the histogram
+    #SPL['all'] = InterpCubicSpline(0.5 * (bins[:-1] + bins[1:]), VALS['all'].astype(float))
+    SPL['all'] = CubicSpline(0.5 * (bins[:-1] + bins[1:]), VALS['all'].astype(float))
+
+    delta_mag2 = delta_mag
+    magbins = np.arange(17, 26, delta_mag2)
+    # magbins = [15,19,21,22,26]
+    rvals = np.linspace(0, zmax, 1e4)
+    for mag in magbins:
+        q = 0
+        while True:
+            cond = (galreal.MAG <= mag + delta_mag2 * 0.5) & (galreal.MAG > mag - delta_mag2 * 0.5)
+            if np.sum(cond) >= Nmin:
+                break
+            else:
+                delta_mag2 += 0.25
+            q += 1
+            assert q < 1000, 'Something wrong with the redshift distribution'
+
+        aux_hist, _ = np.histogram(galreal.ZGAL[cond], bins)
+        VALS['{}'.format(mag)] = gf(aux_hist.astype(float), smooth_scale)  # smooth the histogram
+        SPL['{}'.format(mag)] = CubicSpline(0.5 * (bins[:-1] + bins[1:]), VALS['{}'.format(mag)].astype(float))
+        vals = VALS['{}'.format(mag)]
+        spl = SPL['{}'.format(mag)]
+        rand_z = RanDist(rvals, spl(rvals))
+        if debug:
+            import matplotlib.pyplot as pl
+            pl.plot(bins, spl(bins), '-', label='{}'.format(mag))
+            pl.plot(bins[:-1], aux_hist, drawstyle='steps-mid')
+            pl.xlim(0, 2)
+            pl.legend()
+            pl.show()
+    if debug:
+        import matplotlib.pyplot as pl
+        pl.legend()
+        pl.show()
+
+    for i in range(len(galreal)):
+        if (galreal.MAG[i] > 90) or (galreal.MAG[i] < -90):  # no magnitude, use the whole distribution
```

Comment:
These are also arbitrary to denote "no magnitude" . We should implement something more robust.

Reply:
Added as a TODO.  Am open to suggested ways
to improve..